README.md
[![Bower version](https://badge.fury.io/bo/mm-angular-error-http-handler.svg)](http://badge.fury.io/bo/mm-angular-error-http-handler) [![Build Status](https://secure.travis-ci.org/martinmicunda/mm-angular-error-http-handler.png)](http://travis-ci.org/martinmicunda/mm-angular-error-http-handler) [![Coverage Status](https://coveralls.io/repos/martinmicunda/mm-angular-error-http-handler/badge.png?branch=master)](https://coveralls.io/r/martinmicunda/mm-angular-error-http-handler?branch=master)

 mm-angular-error-http-handler
===================

The `mm.errorHttpHandler` module checks if any [`$http`](https://docs.angularjs.org/api/ng/service/$http) request has failed and if so automatically posts the error to [`Sentry`](https://getsentry.com/) or some custom reporting tool. For any http error you will be redirect to follow pages:

- 400 error status -> /400 (Bad Request)
- 404 error status -> /404 (Not Found)
- 0 error status   -> /500 (Network Error)
- any other error  -> /500 (Internal Server Error)

> **Note:** It's up to you how you gonna handle route in your application and implement UI templates for each error. Have look for simple example in [demo](demo/index.html) and [gulp serve](https://github.com/martinmicunda/mm-angular-error-http-handler/blob/master/gulpfile.js#L243) task.

<br/>

> **Requirements:** AngularJS 1.2+, mm-angular-logger 0.3+ 

> **File Size:** 17.01Kb minified, 1.06Kb gzipped

## Install

Install via bower:
```bash
$ bower install mm-angular-error-http-handler
```

Inject the `mm.errorHttpHandler` module into your app:
```js
var app = angular.module('ErrorHttpDemo', [
    'mm.errorHttpHandler'
]);
```

## Configuration
By default posting error to reporting tool is turn off. This can be turn on by configuring the `errorHttpHandlerProvider`.

### Custom reporting tool:
It's up to you to provide your own back end implementation that accept `post` json request. Have look for simple example in [demo](demo/index.html) and [gulp serve](https://github.com/martinmicunda/mm-angular-error-http-handler/blob/master/gulpfile.js#L243) task.

```js
app.config(['errorHttpHandlerProvider', function (cfg) {
    cfg.setConfigAppErrorPrefix('demoApp'); // optional
    cfg.setConfigCustomDns('http://localhost:3000/error');
}]);
```

### Sentry reporting tool:
You need to create Sentry account to be able to send error data to [`Sentry`](https://getsentry.com/).

```html
<script src="//cdn.ravenjs.com/1.1.15/raven.min.js"></script>
```
```js
app.config(['errorHttpHandlerProvider', function (cfg) {
    var ravenConfig = {logger: 'javascript'};
    cfg.setConfigAppErrorPrefix('demoApp'); // optional
    cfg.setConfigRavenDns('https://7be.......491@app.getsentry.com/2...2');
    cfg.setConfigRavenConfig(ravenConfig); // optional
}]);
```

## Developers
Clone the repo, `git clone git://github.com/martinmicunda/mm-angular-error-http-handler.git`. 

The `mm-angular-error-http-handler` is tested with `karma` against the latest stable release of AngularJS.

```bash
$ npm install
$ gulp test
```
You can build the latest version using `gulp`.
```bash
$ gulp build
```
You can quickly hack around with:
```
$ gulp serve
```

## Release

> During the release process only [`dist`](dist), [`package.json`](package.json), [`bower.json`](bower.json) and [`CHANGELOG.md`](CHANGELOG.md) files should be edited and all steps should be done with **gulp tasks** and **not manually**!

- To push a new release:
  1. Update [`package.json`](package.json) and [`bower.json`](bower.json) version to new version with `gulp bump --type=(major|minor|patch)` 
  2. Build dist files with `gulp build`
  3. Generate changelog with `gulp changelog`
  4. Go through the changelog, and fix any mistakes or clarify any unclear commit messages
  5. Commit and push [`dist`](dist/), [`package.json`](package.json), [`bower.json`](bower.json) and [`CHANGELOG.md`](CHANGELOG.md) with `gulp release` 

## Versioning

Releases will be numbered with the following format:

`<major>.<minor>.<patch>`

And constructed with the following guidelines:

* Breaking backward compatibility bumps the major (and resets the minor and patch)
* New additions without breaking backward compatibility bumps the minor (and resets the patch)
* Bug fixes and misc changes bumps the patch

For more information on SemVer, please visit <http://semver.org/>.

## Contributing
Please submit all pull requests the against master branch. If your code contains JavaScript patches or features, you should include relevant unit tests. See [CONTRIBUTING](CONTRIBUTING.md) file more details. Thanks!

## License

    The MIT License
    
    Copyright (c) 2014 Martin Micunda  

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:
    
    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.
    
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
    THE SOFTWARE.


CHANGELOG.md
### 0.0.2 (2014-08-08)


#### Features

* **400:** handle 400 error ([20d00140](https://github.com/martinmicunda/mm-angular-error-http-handler/commit/20d00140f037674224b4170ef49b50c1513c6cf9))


### 0.0.1 (2014-07-24)

- init release



src/mm-angular-error-http-handler.js
/* global Raven:true */
(function() {
    'use strict';

    /**
     * @ngdoc module
     * @name mm.errorHttpHandler
     *
     * @description
     * The `mm.errorHttpHandler` module checks if any $http request has failed and if so automatically posts the error to {@link https://getsentry.com/ `Sentry`)
     * or some custom reporting tool. For any http error you will be redirect to follow pages:
     *
     * - 400 error status -> /400 (Bad Request)
     * - 404 error status -> /404 (Not Found)
     * - 0 error status   -> /500 (Network Error)
     * - any other error  -> /500 (Internal Server Error)
     * > **Note:** It's up to you how you gonna handle route in your application and implement UI templates for each error.
     *
     * # Usage
     * See {@link mm.errorHttpHandler.errorHttpHandler#usage `errorHttpHandler`} for an example of configuring and using `errorHttpHandler`.
     *
     */

    /**
     * @ngdoc provider
     * @name errorHttpHandlerProvider
     * @module mm.errorHttpHandler
     * @requires ng.$window
     * @requires ng.$injector
     * @requires ng.$q
     * @requires ng.$location
     * @requires mm.Logger
     *
     * @description
     * Use the `errorHttpHandlerProvider` to configure where to post errors. errorHttpHandlerProvider can posts error to `Sentry` or some custom reporting tool
     * see usage below.
     *
     * @usage
     * By default posting error to reporting tool is turn off. This can be turn on by configuring the `errorHttpHandlerProvider` like this (custom reporting tool):
     *
     * ```js
     *   angular.module('customErrorHttpHandlerDemo', []).config(['errorHttpHandlerProvider', function (cfg) {
     *      cfg.setConfigAppErrorPrefix('demoApp');
     *      cfg.setConfigCustomDns('http://localhost:3000/error');
     *   }]);
     * ```
     * > **Note:** You need to provide your own back end implementation that accept `post` json request.
     *
     * or post error to `Sentry` like this:
     *
     * ```html
     * <script src="//cdn.ravenjs.com/1.1.15/raven.min.js"></script>
     * ```
     * ```js
     *   angular.module('sentryErrorHttpHandlerDemo', []).config(['errorHttpHandlerProvider', function (cfg) {
     *      var ravenConfig = {logger: 'javascript'};
     *      cfg.setConfigAppErrorPrefix('demoApp');
     *      cfg.setConfigRavenDns('https://7be...............491@app.getsentry.com/2...2');
     *      cfg.setConfigRavenConfig(ravenConfig);
     *   }]);
     * ```
     * > **Note:** You need to create Sentry account to be able to send error data to {@link https://getsentry.com/ `Sentry`).
     */
    function errorHttpHandlerProvider() {
        /*jshint validthis: true */
        /**
         * @description
         * Error http handler configuration. appErrorPrefix is optional type. If you want to use error reporting tool then at
         * least one of the `ravenDns` or `customDns` has to be defined.
         *
         * @private
         * @type {{appErrorPrefix: string, ravenDns: null, ravenConfig: null, customDns: null}}
         */
        var config = {
            appErrorPrefix: '', // optional
            ravenDns: null,
            ravenConfig: null, // optional
            customDns: null
        };
        /**
         * @ngdoc method
         * @name $exceptionHandlerProvider#setConfigAppErrorPrefix
         *
         * @description
         * Set app error prefix.
         *
         * @param {string=} _appErrorPrefix - an optional app error prefix
         */
        this.setConfigAppErrorPrefix = function(_appErrorPrefix) {
            config.appErrorPrefix = _appErrorPrefix || config.appErrorPrefix;
        };
        /**
         * @ngdoc method
         * @name $exceptionHandlerProvider#setConfigAppErrorPrefix
         *
         * @description
         * Set the raven DNS.
         *
         * @param {string} _ravenDns - an valid raven DNS
         */
        this.setConfigRavenDns = function(_ravenDns) {
            if (!_ravenDns) {
                throw new Error('RavenDNS must be set!');
            }
            config.ravenDns = _ravenDns;
        };
        /**
         * @ngdoc method
         * @name $exceptionHandlerProvider#setConfigRavenConfig
         *
         * @description
         * Set the raven config options.
         *
         * @param {string[]=} _ravenConfig - an optional array of raven config options
         */
        this.setConfigRavenConfig = function(_ravenConfig) {
            config.ravenConfig = _ravenConfig || config.ravenConfig;
        };
        /**
         * @ngdoc method
         * @name $exceptionHandlerProvider#setConfigCustomDns
         *
         * @description
         * Set the custom dns.
         *
         * @param {string} _customDns - an valid custom DNS
         */
        this.setConfigCustomDns = function(_customDns) {
            if (!_customDns) {
                throw new Error('CustomeDNS must be set!');
            }
            config.customDns = _customDns;
        };

        this.$get = ['$injector', '$q', '$location', '$window', 'Logger', function ($injector, $q, $location, $window, Logger) {
            var logger = Logger.getLogger('$exceptionHandler');

            var browserInfo = {
                navigatorAppName : navigator.appName,
                navigatorUserAgent : navigator.userAgent
            };
            if ($window.Raven) {
                logger.info('Using the RavenJS server-side logging exception handler.');
                logger.info('RavenJS ravenDns {0}', [config.ravenDns]);
                logger.info('RavenJS ravenConfig {0}', [config.ravenConfig]);
                try {
                    // configure Raven
                    Raven.config(config.ravenDns, config.ravenConfig).install();
                } catch (loggingError) {
                    logger.error('Error to configure RavenJS');
                    logger.error(loggingError.toString());
                }
            } else if(config.customDns) {
                logger.info('Using the custom server-side logging exception handler.');
                logger.info('Custom DNS {0}', [config.customDns]);
            }

            return {
                responseError: function (rejection) {
                    try {
                        if ($window.Raven) {
                            // post error to Sentry
                            Raven.captureException(new Error('HTTP response error'), {extra: {
                                method: rejection.config.method,
                                url: rejection.config.url,
                                message: rejection.data,
                                status: rejection.status,
                                browserInfo: browserInfo}
                            });
                        } else if(config.customDns) {
                            // using the $injector service directly to avoid a circular dependency
                            var $httpBackend = $injector.get('$httpBackend');

                            // bypassing angular's http abstraction to avoid infinite exception loops
                            $httpBackend('POST', config.customDns, angular.toJson({
                                method: rejection.config.method,
                                url: rejection.config.url,
                                message: rejection.data,
                                status: rejection.status,
                                browserInfo: browserInfo
                            }), angular.noop, { 'content-type': 'application/json' });
                        }
                    } catch (loggingError){
                        logger.error('Error server-side logging failed');
                        logger.error(loggingError.toString());
                    }
                    switch (rejection.status) {
                        case 400:
                            $location.path('/400');
                            logger.error('{0} {1} \'{2}\'', [rejection.status, rejection.statusText, rejection.config.url]);
                            break;                        
                        case 404:
                            $location.path('/404');
                            logger.error('{0} {1} \'{2}\'', [rejection.status, rejection.statusText, rejection.config.url]);
                            break;
                        case 0:
                            // TODO (martin): this error should be custom error page and not 500
                            $location.path('/500');
                            logger.error('{0} {1} {2} \'{3}\'', [rejection.status, rejection.statusText, rejection.config.url, 'Unable to communicate with the server. Make sure you are connected to the internet and try again.']);
                            break;
                        default:
                            $location.path('/500');
                            logger.error('{0} {1} {2} \'{3}\'', [rejection.status, rejection.statusText, rejection.config.url, rejection.data.message]);
                            break;
                    }
                    return $q.reject(rejection);
                }
            };
        }];
    }

    angular
        .module('mm.errorHttpHandler', ['mm.logger'])
        .provider('errorHttpHandler', errorHttpHandlerProvider)
        .config(function ($httpProvider) {
            $httpProvider.interceptors.push('errorHttpHandler');
        });
})();


gulpfile.js
'use strict';

//=============================================
//            DECLARE CORE VARIABLES
//=============================================

/**
 * Load required core dependencies. These are installed based on the versions listed
 * in 'package.json' when you do 'npm install' in this directory.
 */
var fs          = require('fs');
var pkg         = require('./package.json');
var open        = require("open");
var path        = require('path');
var argv        = require('minimist')(process.argv.slice(2));
var gulp        = require('gulp');
var karma       = require('karma').server;
var semver      = require('semver');
var browser     = require('tiny-lr')();
var express     = require('express');
var wiredep     = require('wiredep').stream;
var changelog   = require('conventional-changelog');
var runSequence = require('run-sequence');


//=============================================
//            DECLARE GULP VARIABLES
//=============================================

/**
 * Load required Gulp tasks. These are installed based on the versions listed
 * in 'package.json' when you do 'npm install' in this directory.
 */
var size                 = require('gulp-size');
var gzip                 = require('gulp-gzip');
var bump                 = require('gulp-bump');
var help                 = require('gulp-help')(gulp);
var exec                 = require('gulp-exec');
var gutil                = require('gulp-util');
var bower                = require('gulp-bower');
var watch                = require('gulp-watch');
var clean                = require('gulp-clean');
var gulpif               = require('gulp-if');
var jshint               = require('gulp-jshint');
var header               = require('gulp-header');
var uglify               = require('gulp-uglify');
var rename               = require('gulp-rename');
var concat               = require('gulp-concat');
var refresh              = require('gulp-livereload');
var sourcemaps           = require('gulp-sourcemaps');
var ngAnnotate           = require('gulp-ng-annotate');
var coverageEnforcer     = require("gulp-istanbul-enforcer");


//=============================================
//            DECLARE CONSTANTS
//=============================================

/**
 * Declare constants that are use in gulpfile.js or angular app
 */
var COLORS               = gutil.colors;
var DEMO_PORT            = 9000;
var LIVERELOAD_PORT      = 35729;

//=============================================
//            DECLARE VARIABLES
//=============================================
/**
 * Declare variables that are use in gulpfile.js
 */
var hasGitChanges   = '';
var isWatching      = false;


//=============================================
//            DECLARE PATHS
//=============================================

var paths = {
    gulpfile:       'gulpfile.js',
    dist:           'dist/',
    src:            'src/*.js',
    vendor:         'bower_components/',
    html:           'demo/*.html',
    tmp:            '.tmp',
    testReports:    '.tmp/test-reports/coverage/'
};


//=============================================
//            DECLARE BANNER DETAILS
//=============================================

/**
 * The banner is the comment that is placed at the top of our compiled
 * source files. It is first processed as a Grunt template, where the `<%=`
 * pairs are evaluated based on this very configuration object.
 */
var banner = gutil.template('/**\n' +
    ' * <%= pkg.description %>\n' +
    ' * @version v<%= pkg.version %> - <%= today %>\n' +
    ' * @link <%= pkg.homepage %>\n' +
    ' * @author <%= pkg.author.name %>\n' +
    ' * @copyright <%= year %>(c) <%= pkg.author.name %>\n' +
    ' * @license <%= pkg.licenses.type %>, <%= pkg.licenses.url %>\n' +
    ' */\n', {file: '', pkg: pkg, today: new Date().toISOString().substr(0, 10), year: new Date().toISOString().substr(0, 4)});


//=============================================
//               SUB TASKS
//=============================================

gulp.task('clean', 'Delete \'dist\' directory', function () {
    return gulp.src([paths.dist, paths.tmp], {read: false})
        .pipe(clean());
});

gulp.task('copy', 'Copy no-minified source files into \'dist\' folder.', function () {
    return gulp.src(paths.src)
        .pipe(concat(pkg.name + '.js'))
        .pipe(header(banner, { pkg : pkg, date: new Date }))
        .pipe(gulp.dest(paths.dist));
});

gulp.task('bower', 'Install all bower dependencies specify in bower.json from bower repository', function () {
    return bower();
});

gulp.task('bower-install', 'Does the same as \'bower\' task but also inject bower components to index.html', ['bower'], function () {
    return gulp.src(paths.html)
        .pipe(wiredep({
            directory: paths.vendor
        }))
        .pipe(gulp.dest('./demo'))
        .pipe(size());
});

gulp.task('jshint', 'Hint src JavaScripts files', function () {
    return gulp.src(paths.src)
        .pipe(jshint('.jshintrc'))
        .pipe(jshint.reporter('jshint-stylish'))
        .pipe(gulpif(!isWatching, jshint.reporter('fail')))
        .pipe(gulpif(isWatching, refresh(browser)))
        .pipe(size());
});

gulp.task('compress', 'Compress js files to min.js', function() {
    return gulp.src(paths.src)
        .pipe(sourcemaps.init())
        .pipe(concat(pkg.name + '.js'))
        .pipe(ngAnnotate())
        .pipe(uglify())
        .pipe(sourcemaps.write())
        .pipe(header(banner))
        .pipe(rename({suffix: '.min'}))
        .pipe(gulp.dest(paths.dist))
        .pipe(size())
        .on('error', function (error) {
            gutil.log(COLORS.red('Error: compress failed '  + error));
            return process.exit(1);
        });
});

gulp.task('compress:gzip', 'Compress js files to min.js.gz', function() {
    return gulp.src(paths.src)
        .pipe(concat(pkg.name + '.js'))
        .pipe(ngAnnotate())
        .pipe(uglify())
        .pipe(header(banner))
        .pipe(rename({suffix: '.min'}))
        .pipe(gulp.dest(paths.dist))
        .pipe(gzip())
        .pipe(gulp.dest(paths.dist))
        .pipe(size())
        .on('error', function (error) {
            gutil.log(COLORS.red('Error: compress:zip failed ' + error));
            return process.exit(1);
        });
});

gulp.task('watch', 'Watch files for changes', function () {

    // set to 'true' to avoid process exit on error for jshint
    isWatching = true;

    // Listen on port 35729
    browser.listen(LIVERELOAD_PORT, function (err) {
        if (err) {
            return console.log(err);
        }

        gulp.watch([
            paths.html
        ], function (event) {
            return gulp.src(event.path)
                .pipe(refresh(browser));
        });

        // Watch js files
        gulp.watch(paths.src, ['jshint']);

        // Watch bower file
        gulp.watch('bower.json', ['bower-install']);
    });

});

gulp.task('karma', 'Run karma unit tests', ['bower-install'], function (cb) {
    // remove 'coverage' directory before each test
    gulp.src(paths.testReports, {read: false})
        .pipe(clean())
        .on('finish', function() {
            // run the karma test
            karma.start({
                configFile: path.join(__dirname, 'karma.conf.js'),
                browsers: ['PhantomJS']
            }, function(code) {
                // make sure failed karma tests cause gulp to exit non-zero
                if(code === 1) {
                    gutil.log(COLORS.red('Error: unit test failed '));
                    process.exit(1);
                }
                cb();
            });
        });
});

gulp.task('check', 'Check if there are any changes to commit', function (cb) {
    require('child_process').exec('git status --porcelain', function (err, stdout) {
        hasGitChanges = stdout;
        cb(err);
    });
});

//=============================================
//                MAIN TASKS
//=============================================

/**
 * The demo web server.
 */
gulp.task('serve', 'Run the demo web server', ['bower-install', 'watch'], function() {
    var app = express();
    app.use(express.static('./'));
    app.use(express.static('demo'));
    app.listen(DEMO_PORT);
    app.post('/error', function(req, res) {
        console.log("Got UI error:\n", JSON.stringify(req.body, null, 2));
        res.json(200);
    });
    app.get('/error404', function(req, res) {
        console.log('return 404 error');
        res.status(404);
        res.send('throw 404 error');
    });
    app.get('/error500', function(req, res) {
        console.log('return 500 error');
        res.status(500);
        res.send('throw 500 error');
    })
    app.get('*', function(req, res, next) {
        next();
    });
    if(gutil.env.open) {
        gutil.log('Opening dev server URL in browser');
        open('http://localhost:' + DEMO_PORT);
    } else {
        gutil.log(COLORS.yellow('Hint: Run with --open to automatically open URL on startup'));
    }
});

/**
 * Run unit test and coverage.
 */
gulp.task('test', 'Run unit tests and coverage', ['karma'], function () {
    var options = {
        thresholds : {
            statements : 95,
            branches : 90,
            lines : 95,
            functions : 95
        },
        coverageDirectory: paths.testReports,
        rootDirectory : paths.tmp // keep root `.tmp` so enforce plugin is not searching in other directories
    };
    return gulp.src('.')
//        .pipe(coverageEnforcer(options))
        .on('error', function(error) {
            gutil.log(error.toString());
            process.exit(1);
        });
});

/**
 * The 'build' task gets app ready for deployment by
 * minifying etc.
 */
gulp.task('build', 'Build dist files', function (cb) {
    runSequence(
        ['clean', 'jshint', 'bower-install'],
        'test',
        ['compress', 'compress:gzip', 'copy'],
        cb);
});

/**
 * Bump version number in package.json & bower.json.
 */
gulp.task('bump', 'Bump version number in package.json & bower.json', ['jshint'], function () {
    var HAS_REQUIRED_ATTRIBUTE = !!argv.type ? !!argv.type.match(new RegExp(/major|minor|patch/)) : false;

    if (!HAS_REQUIRED_ATTRIBUTE) {
        gutil.log(COLORS.red('Error: Required bump \'type\' is missing! Usage: gulp bump --type=(major|minor|patch)'));
        return process.exit(1);
    }

    if (!semver.valid(pkg.version)) {
        gutil.log(COLORS.red('Error: Invalid version number - ' + pkg.version));
        return process.exit(1);
    }

    return gulp.src(['package.json', 'bower.json'])
        .pipe(bump({type: argv.type}))
        .pipe(gulp.dest('./'));
});

/**
 * Generate changelog.
 */
gulp.task('changelog', 'Generate changelog', function(callback) {
    changelog({
        version: pkg.version,
        repository: 'https://github.com/martinmicunda/' + pkg.name
    }, function(err, data) {
        if (err) {
            gutil.log(COLORS.red('Error: Failed to generate changelog ' + err));
            return process.exit(1);
        }
        fs.writeFileSync('CHANGELOG.md', data, callback());
    });
});

/**
 * The 'release' task push package.json and CHANGELOG.md to GitHub.
 */
gulp.task('release', 'Release bumped version number to GitHub repo', ['check'], function () {
    if (!semver.valid(pkg.version)) {
        gutil.log(COLORS.red('Error: Invalid version number - ' + pkg.version + '. Please fix the the version number in package.json and run \'gulp publish\' command again.'));
        return process.exit(1);
    }

    if (hasGitChanges === '') {
        gutil.log(COLORS.red('Error: No changes detected in this repo. Aborting release.'));
        return process.exit(1);
    }

    gutil.log(COLORS.blue('Pushing to GitHub ...'));
    var commitMsg = 'chore(release): v' + pkg.version;
    return gulp.src('package.json')
        .pipe(exec('git add dist CHANGELOG.md package.json bower.json'))
        .pipe(exec('git commit -m "' + commitMsg + '" --no-verify'))
        .pipe(exec('git push origin master'));
});
